prometheus-proc/src/Prometheus/Metric/Proc.hs

```diff
@@ -0,0 +1,175 @@
+{-# language ForeignFunctionInterface #-}
+{-# language OverloadedStrings #-}
+{-# language RecordWildCards #-}
+
+module Prometheus.Metric.Proc ( procMetrics ) where
+
+import Data.Char ( isSpace )
+import Data.Maybe ( catMaybes )
+import Data.String ( fromString )
+import Foreign.C
+import Prometheus
+import System.FilePath
+import System.IO.Unsafe
+import System.Posix.Memory ( sysconfPageSize )
+import System.Posix.Process ( getProcessID )
+import qualified Text.Regex.Applicative as RE
+import qualified Text.Regex.Applicative.Common as RE
+
+
+data ProcMetrics =
+  ProcMetrics
+
+
+procMetrics :: Prometheus.Metric ProcMetrics
+procMetrics =
+  Metric ( return ( ProcMetrics, collect ) )
+
+
+foreign import ccall unsafe
+  clk_tck :: CLong
+
+
+collect :: IO [ SampleGroup ]
+collect = do
+  pid <-
+    getProcessID
+
+  mprocStat <-
+      RE.match parseProcStat <$> readFile ( "/" </> "proc" </> show pid </> "stat" )
+
+  return
+    ( foldMap ( toMetrics ) mprocStat )
+
+
+toMetrics :: ProcStat -> [ SampleGroup ]
+toMetrics ProcStat{..} =
+  catMaybes
+    [ Just process_cpu_seconds_total
+    , process_start_time_seconds
+    , Just process_virtual_memory_bytes
+    , Just process_resident_memory_bytes
+    ]
+
+  where
+
+    process_cpu_seconds_total =
+      metric
+        "process_cpu_seconds_total"
+        "Total user and system CPU time spent in seconds."
+        CounterType
+        ( fromTicks ( utime + stime ) )
+
+    process_start_time_seconds = do
+      btime <-
+        mbtime
+
+      return
+        ( metric
+            "process_start_time_seconds"
+            "Start time of the process since unix epoch in seconds."
+            GaugeType
+            ( fromIntegral btime + fromTicks starttime )
+        )
+
+    process_virtual_memory_bytes =
+      metric
+        "process_virtual_memory_bytes"
+        "Virtual memory size in bytes."
+        GaugeType
+        vsize
+
+    process_resident_memory_bytes =
+      metric
+        "process_resident_memory_bytes"
+        "Resident memory size in bytes."
+        GaugeType
+        ( rss * sysconfPageSize )
+
+    fromTicks ticks =
+      fromIntegral ticks / fromIntegral clk_tck :: Double
+
+    metric metricName metricHelp metricType value =
+      SampleGroup
+        Info{..}
+        metricType
+        [ Sample
+            metricName
+            []
+            ( fromString ( show value ) )
+        ]
+
+
+{-# NOINLINE mbtime #-}
+mbtime :: Maybe Int
+mbtime = unsafePerformIO $ do
+  fmap ( \( _, a, _ ) -> a ) . RE.findFirstInfix ( "btime " *> RE.decimal )
+    <$> readFile "/proc/stat"
+
+
+data ProcStat = ProcStat
+  { utime :: Int
+  , stime :: Int
+  , starttime :: Int
+  , vsize :: Int
+  , rss :: Int
+  }
+  deriving
+    ( Show )
+
+
+parseProcStat :: RE.RE Char ProcStat
```

Comment:
Why did you lean to a regex library? I would have reached for Attoparsec. No judgment, just curious.

Reply:
I guess two reasons:

1. `attoparsec` is quite a big hammer - the language of `/proc/xxx/stat` is not context sensitive.
2. The process name field is actually quite subtle to parse. It's the name, surrounded by brackets, but otherwise unescaped. A process name *could* contain a bracket, so just looking for the next `)` isn't sufficient. However this falls out quite naturally when using a regular expression - just chomp as many characters as you want until you get to `)` (greedy matching), and then parse the rest of the fields. The correct solution just falls out of this, with no need to be explicit about backtracking.